posts/2021-02-26-const-generics-mvp-beta.md

```diff
@@ -0,0 +1,127 @@
+    // ...
+}
+
+let arr = [vec![0, 1], vec![1, 2, 3], vec![3]];
+for elem in array::IntoIter::new(arr) {
+    needs_vec(elem);
+}
+```
+
+## What's next?
+
+### Const generics and default arguments
+
+Generic parameters must currently come in a specific order: lifetimes, types, consts. However, this causes difficulties when one attempts to use default arguments alongside const parameters. For the compiler to know which generic argument is which, any default arguments need to be placed last. These two constraints - "types come before consts", and "defaults come last" - conflict with each other for definitions that have default type arguments *and* const parameters.
+
+The solution to this is to relax the ordering constraint so that const parameters may precede type arguments. However, there turn out to be subtleties involved in implementing this change, because the Rust compiler currently makes assumptions about parameter ordering that require some delicacy to remove.
+
+In light of similar design questions around defaults for const arguments, these are also currently not supported in version 1.51. However, fixing the parameter ordering issues above will also unblock const defaults.
+
+### Const generics for custom types
+
+For a type to be valid, in theory, as the type of a const parameter, we must be able to compare values of that type at compile-time. Furthermore, equality of values should be well-behaved (namely, it should be deterministic, reflexive, symmetric, and transitive). To guarantee these properties, the concept of *structural equality* was introduced in the [const generics RFC](https://github.com/rust-lang/rfcs/blob/master/text/2000-const-generics.md): essentially this includes any type with `#[derive(PartialEq, Eq)]` whose members also satisfy structural equality.
+
+There are [still some questions](https://github.com/rust-lang/rust/issues/74446) concerning precisely how structural equality should behave, and [prerequisites for implementation](https://github.com/rust-lang/compiler-team/issues/323). Primitive types are significantly simpler, which has allowed use to stabilize const generics for these types before more general types.
+
+### Const generics with complex expressions
+
+There are several complexities involved in supporting complex expressions. A feature flag, `feature(const_evaluatable_checked)`, is available in the Nightly channel, which enables a version of complex expression support for const generics.
+
+One difficulty lies in the necessity of having some way to compare unevaluated constants, as the compiler does not automatically know that two syntactically identical expressions are actually equal. This involves a kind of symbolic reasoning about expressions, which is a complex problem in general.
+```rust
+// The two expressions `N + 1` and `N + 1` are distinct
+// entities in the compiler, so we need a way to check
+// if they should be considered equal.
+fn foo<const N: usize>() -> [u8; N + 1] {
+    [u8; N + 1]
```

Comment:
This should be `[0; N + 1]` or similar.

Reply:
This was fixed in https://github.com/rust-lang/blog.rust-lang.org/pull/785.

Reply:
Ah, perfect :)